### PR TITLE
[DCP][Quantization] Fix for FP8 multiplication during dequantization

### PR DESCRIPTION
**Pull request by saumishr**
Summary:
Weight vector needs to be upcasted since some FP8 formats (like Float8_e4m3fn) don't have CPU implementations in PyTorch. Reference: https://docs.pytorch.org/docs/stable/tensors.html#id13

We will use FP32 for the scale vector multiplication and convert to the target dtype. 

Upcasting helps with the following:

1.  **Full CPU support**: `float32` has complete CPU kernel implementations for all operations
2.  **Numerical stability**: `float32` provides more precision during intermediate calculations
3.  **Compatibility**: Works across all devices (CPU/GPU) and PyTorch versions

Test Plan:
UTs

Rollback Plan:

Differential Revision: D81711093

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim

### CODE REVIEW

**Review by wwwjn — APPROVED** [review]

File: torch/distributed/checkpoint/quantized_hf_storage.py [hunk 1]
```diff
@@ -131,7 +136,7 @@ def _dequantize_tensor(
         block_rows, block_cols = expected_scale_shape
 
         # Create output tensor in target dtype
-        dequantized = weight.detach().clone().to(dtype=self.target_dtype)
+        dequantized = upcasted_weight.detach().clone().to(dtype=self.target_dtype)
```
Comment (vadimkantorov) [inline torch/distributed/checkpoint/quantized_hf_storage.py]: is it the same as `upcasted_weight.detach().to(dtype=self.target_dtype, copy=True)`?

Comment (saumishr) [reply torch/distributed/checkpoint/quantized_hf_storage.py]: Yes. I believe so. In fact, I feel your suggestion may be more efficient. I will verify that and update the PR.

Comment (vadimkantorov) [reply torch/distributed/checkpoint/quantized_hf_storage.py]: I also wonder if typically it would be faster to do `dequantized = weight.detach().to(dtype=self.target_dtype, copy=True)` - if `weight` is typically lower-width dtype than float32, it could spent less bandwidth/cycles reading input `weight` from VRAM (than reading `upcasted_weight` which float32)

Comment (saumishr) [reply torch/distributed/checkpoint/quantized_hf_storage.py]: @vadimkantorov Great suggestion. We can probably work with the quantized weight here. For larger tensors, it may be more efficient. I will update the PR.
